jupyter-js-widgets/src/widget.ts

```diff
@@ -399,23 +406,82 @@ class WidgetModel extends Backbone.Model {
             // get binary values, then send
             var keys = Object.keys(state);
             var buffers = [];
-            var buffer_keys = [];
-            for (var i=0; i<keys.length; i++) {
-                var key = keys[i];
-                var value = state[key];
-                if (value) {
-                    if (value.buffer instanceof ArrayBuffer
-                        || value instanceof ArrayBuffer) {
-                        buffers.push(value);
-                        buffer_keys.push(key);
-                        delete state[key];
+            var buffer_paths = [];
+            // keep track of what we visited, to avoid endless loops, when an object ancester
+            // refers to itself, we cannot use an object, since using objects for keys don't work
+            var visited_set = [];
+            // this function goes through lists and object and removes arraybuffers
+            // they will be transferred seperately, since they don't json'ify
+            // on the python side the inverse happens
+            // if we need to remove an object from a list, we need to clone that list, otherwise we may modify
+            // the internal state of the widget model
+            // however, we do not want to clone everything, for performance
+            function seperate_buffers(obj, path) {
+                if(visited_set.indexOf(obj) != -1) { // we already visited this object
+                    return obj;
+                }
+                visited_set.push(obj);
+                if (_.isArray(obj)) {
```

Comment:
I think all of our supported browsers support [Array.isArray](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/isArray)